### PR TITLE
Add contentMatch to sprites

### PR DESCRIPTION
**Pull request by DavidBuerer**
This is to provide a way to work around shields with small text getting too small (left) without creating multiple layers with multiple copies of the shields.

image

The content rectangle alone is not sufficient to keep shields looking nice when using "icon-text-fit" as it allows the shields to get too narrow if the text is small.  To solve this, a limit is needed on how much distortion can happen to a sprite with content.

A few considerations:
1. Since the size of the text is not known to the containing sprite, a minimum width and height would not work, but an aspect ratio does.
2. We could let the designer specify an aspect ratio, but this would generally require an icon designer to calculate the aspect ratio from their "content" rectangle. Rather than making them do this, I use the "content" rectangle's aspect ratio.
3. In addition to an aspect ratio, we need to know whether we are locking the text size horizontally or vertically to the aspect ratio. A common case will be horizontal for text that is written horizontally, however, if using vertical Chinese characters, for example, vertical should be supported.
4. Since this value could vary per sprite, it is best to make it available on sprite rather than on the layer.

So what I did is add a contentMatch enum to sprites that may be 'none', 'height', and 'width'.  The content rectangle will stretch to match the specified dimension and the other dimension will scale with the content, but not shrink below the aspect ratio of the content rectangle.  

This image illustrates the effect that the different values have on sprites.

image

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [x] Manually test the debug page.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

**Comment by HarelM** [comment]
Thanks for taking the time and open this PR!
A few little comments:
1. It would be great to add unit tests as well besides the render tests (which are great)
2. I don't understand the terminology and how would one use it, it means that the definition of the feature is not clear and the wording around it need some improvement. I.e what does contentMatch: width mean? I looked at the picture a few time and still it's not clear to me what I should expect when setting this value...

**Comment by DavidBuerer** [comment]
> Thanks for taking the time and open this PR! A few little comments:
> 
> 1. It would be great to add unit tests as well besides the render tests (which are great)
> 2. I don't understand the terminology and how would one use it, it means that the definition of the feature is not clear and the wording around it need some improvement. I.e what does contentMatch: width mean? I looked at the picture a few time and still it's not clear to me what I should expect when setting this value...

Render tests exercise the code in question, so I'm not sure of the value of additional unit tests, but I'll take a look.

It is a little tough to express the concept in a single sentence, but maybe the comment in map.ts helps?  I'm open to suggestions on clarifying.  

> options.contentMatch Optional enum value (or `none`) that may match the content's `height` or `width` and limit the other dimension from getting smaller than the aspect ratio of content.

**Comment by HarelM** [comment]
Ok, I think I got it, but it's too complicated... :-(
none=both basically, right?
BTW, cant the different padding solve this (only left and right padding in your example)?

**Comment by DavidBuerer** [comment]
* 'none' is the same as not including contentMatch... which means that the sprite will scale so that the content rectangle will precisely match the content (as it does now).
* 'height' means that the sprite will scale so that the height of the content rectangle will precisely match the content, and the width will try to scale down to match the content, but won't let the content rectangle get narrower than its aspect ratio.
* 'width' is the same as 'height except that it precisely matches the content width and limits the heigh to the aspect ratio.

**Comment by DavidBuerer** [comment]
The icon designer already has to include a content rectangle in their sprite definition and would now need to include a contentMatch='height' value if they want this functionality for a horizontal string.

Sizing the content rectangle to match the content can't preserve the aspect ratio because if you had a long string, for example, the containing sprite would become huge instead of wide.

This feature doesn't preserve the aspect ratio, it stops the content rectangle from becoming smaller than the aspect ratio (using the dimension you specify) so it doesn't get too squished.  It can become as wide (or tall) as you want.

**Comment by wipfli** [comment]
Thanks for the inputs, @zekefarwell. I think we should avoid changing thd behavior of existing style properties.

**Comment by DavidBuerer** [comment]
This doesn't change the behavior.  If you leave out contentMatch, it acts exactly as it does today.  It is only if that property is included (and set to 'width' or 'height') that it changes the behavior.

**Comment by zekefarwell** [comment]
> The icon designer already has to include a content rectangle in their sprite definition

You don't have to include a content rectangle in the [sprite definition](https://maplibre.org/maplibre-gl-js-docs/style-spec/sprite/).  The `content` property is optional.

> would now need to include a contentMatch='height' value if they want this functionality for a horizontal string.

Ok so it sounds like the answer to my question about how a style author would use this feature is that they would add the contentMatch property to the sprite definition in the sprite index JSON file as shown below.  This was not clear to me from the PR description.

```
{
    "poi": {
        "width": 32,
        "height": 32,
        "x": 0,
        "y": 0,
        "pixelRatio": 1,
        "contentMatch": "height",
    }
}
```

I think I somewhat understand how this is intended to work, but it's unintuitive to me.  What would make a lot of sense to me would be to add a new `preserve-aspect-ratio` value to the [icon-text-fit layout property](https://maplibre.org/maplibre-gl-js-docs/style-spec/layers/#layout-symbol-icon-text-fit).  This property would scale the icon so that it fits the text while preserving the icon's aspect ratio (scale both dimensions proportionally).  However, since this PR specifies separate height and width values, maybe this isn't quite the same as what you're aiming for.  It seems like you do want to allow the aspect ratio of the icon to change, but not too much or something?  It's a bit confusing to me.

**Comment by DavidBuerer** [comment]
This feature does not preserve the aspect ratio.  If the aspect ratio was preserved, then if you put in a long string, the sprite would become large vertically and horizontally trying to preserve it.  What it does is to not let the content rectangle get smaller than the aspect ratio on one of the axis.  This way the sprite can stretch along the axis as it does today, but not deform if the content gets too small.

The reason that I put this on the sprite (#4) is that some sprites in a layer may need it and others may not.  It is undesirable to have to split into multiple layers because you want the content rectangle to behave differently.

**Comment by HarelM** [comment]
@DavidBuerer can you find the CSS equivalent to this feature? It took me a few reading to fully understand it, and the complexity of it (mainly what it should and shouldn't do, not the code itself) will probably cause a low adoption rate (I believe). So I'm a bit hesitated to pull this in.

**Comment by DavidBuerer** [comment]
@HarelM I am unaware of an equivalent, or I would have used it.  I believe the problem attempting to be solved is somewhat unique to scaling an image to fit a content rectangle which, like stretchX/Y are somewhat unique to needing to stretch road shields to fit content in a reasonable way.  I'm open to alterative solutions to the original problem.

**Comment by zekefarwell** [comment]
Apologies for misunderstanding the intent here and confusing the issue further with discussion of aspect ratio preservation.  It does seems there is a good idea here, but I have to agree with @HarelM, that it needs a clearer explaination of how it is supposed to function.

**Comment by HarelM** [comment]
@DavidBuerer can't padding help here? I know it's not the same, but you can define a padding that will cause the shield to look Ok with a single digit and not that big with multiple digits, can't you?

**Comment by DavidBuerer** [comment]
@HarelM If there is padding, then longer strings will also show that padding and look too fat.  In theory you could modify short strings with spaces, but that ties styling directly to data.  What has been done is to make multiple copies of each shield with different content rectangles, but this causes multiple shield sprite copies per shield which affects startup time.

**Comment by HarelM** [comment]
You might be able to add the spaces in the style using some string manipulations, wouldn't you?
Regarding padding, yes, it will cause long text to get it as well, but it should be less noticeable, hopefully, in long text.

**Comment by DavidBuerer** [comment]
Having to modify the incoming data with special rules to compensate for styling is a slippery slope.  I'm not sure I could do justice to the full argument why, but some of the reasons are that any new data source would have to follow those rules, designers would now have to be in the business of fixing data rules, etc.

It is not less noticeable.  Top right is what we want when more text is added.  What you get if you leave the padding in is the bottom right.  It might be possible to squeeze and tuck until the results are only kind of bad, but why?
image

**Comment by HarelM** [comment]
I'm not sure I'm fully aligned with the slippery slope argument. I agree that the shields with padding and long text are not looking as good as they should but, as I said, introducing this feature, which is very hard to understand just in order to solve this particular case is also problematic, from my point of view.
What I believe might help solve this argument is either a more simple approach, maybe something like min-width (without taking into consideration the aspect ratio, as it complicates things a lot) or find more use cases in other places to prove this isn't a specific "patch" for this problem.

**Comment by DavidBuerer** [comment]
min-width doesn't solve the problem because you don't know the width you're dealing with.  The original content width may be 500px but the sprite doesn't know how big the content it will be scaled to (could be up to 1000px or down to 10px).  There is no single width value that works.  That is why preserving the aspect ratio is the best path.

**Comment by HarelM** [comment]
I've added this PR in the coming technical steering committee meeting agenda. You are welcome to join and present it there (it's this Wednesday)

**Comment by DavidBuerer** [comment]
@HarelM, how do I join the steering committee meeting?

**Comment by HarelM** [comment]
It's every second Wednesday of the month. Reminder is sent in the slack channel prior to the meeting.

**Comment by lseelenbinder** [comment]
As discussed the TSC: https://docs.mapbox.com/mapbox-gl-js/style-spec/sprite/#index-file could be a better way to implement the API.

**Comment by nyurik** [comment]
Note to all: please do not even look at the Mapbox implementation code for this feature. We can only read the public documentation.

**Comment by drwestco** [comment]
The `content`, `stretchX` and `stretchY` properties are already in MapLibre. These don't preserve the aspect ratio when shrinking the way this new `contentMatch` property does.

**Comment by HarelM** [comment]
Just to summarize the discussion a bit differently: The main problem right now is the definition of the property name and how to make it more intuitive to use.
Once we do that and have a more intuitive way to define this behavior this can be accepted into maplibre.

**Comment by lseelenbinder** [comment]
@drwestco Oh! I misunderstood on the call this evening then. Carry on. :smile:

**Comment by DavidBuerer** [comment]
Apologies for not making the steering committee meeting, but I was sick.  I went through a number of rounds of naming here.  I will say that this feature is necessary for the performance of appropriately drawn shields but is also fairly esoteric, so there may not be a blatantly obvious naming convention.  But here is another alternative.

The idea is to explain how the content rectangle is limited when it is scaled due to icon-text-fit, so instead of naming it based on the axis that is matched, we could name it based on the axis that is limited:

```
contentFitLimit: {
       none
       xByAspectRatio
       yByAspectRatio
       byAspectRatio
}
```

Keep in mind that this needs to fit in with the sprite definition.  Does this make more sense to people?

```
"nine-part-content-match-height": {
    "x": 0,
    "y": 0,
    "width": 40,
    "height": 40,
    "pixelRatio": 2,
    "stretchX": [[10, 30]],
    "stretchY": [[10, 30]],
    "sdf": false,
    "content": [10, 10, 30, 30],
    "contentFitLimit": "xByAspectRatio"
},
```

Updated usage diagram:

image

**Comment by HarelM** [comment]
I think that using the term aspect ratio, when you are not keeping the aspect ratio (as most people are familiar with what it means to keep an aspect ratio) will create confusion.
I would consider something along min-something, max-something, and maybe limiting it to when the icon is in its original size. and maybe adding a different property to let the reader know that this property can change based on scaling.
for example, and I'm not sure this is better, but I'm brain storming here:
`min-width: 10px`
`min-width-scale: true`
So basically min width will be kept at 10px always if the `min-width-scale` is `false` and it will scale according to how the icon is scaled, so if the icon is 100px wide, and it was scaled down to 50px the min-width will be 5px.
I don't know, it's just a thought, I just prefer to keep the term aspect ratio out of this property as it doesn't really keep the aspect ratio...
Comments are welcomed :-)

**Comment by DavidBuerer** [comment]
But it does use the aspect ratio.  It doesn't let you distort below the aspect ratio.  The reason I used aspect ratio is under an assumption that the original layout of the sprite is probably closest to its desired size, so the value of the content aspect ratio is already known without having to introduce new values that a designer would need to calculate.  Also, I thought people would find it confusing to specify a minimum width that was not actually the on-screen width.  Being able to specify the unscaled width is sort of a different feature, but seems like it could be helpful.  

Note that this is not in the style (hence the lack of kebab-casing) because this is tied to the sprite, not to the layer.

**Comment by HarelM** [comment]
The entire dimensions of the icon on screen might be different depending on a lot of other factor, nevertheless, the units in the sprite.json are all in pixels and thus make sense to use the same units in the same place.
I can't stress this enough: the usage of "aspect ratio" in this context is bad, IMHO.
Another calculation on top of the over complicated icon properties (that feels like were added only for the sake of shields) is a very week argument, sorry...
If I'm looking at this page for help: https://docs.mapbox.com/mapbox-gl-js/style-spec/sprite/ and I need to calculate the `content`, the `stretchX` and `stretchY` calculating the minimal width or height of this icon should be easy...
Bottom line, this is what I have to offer.
If anyone else wants to help out with a different name feel free. :-)

**Comment by wipfli** [comment]
We had a similar discussion in #716 when I added the new property to rotate glyphs to the viewport. I think it is just really hard to come up with good names for these thing.

@DavidBuerer for me it would help a lot if you could add a debug page showcasing the new feature. You could have a look at the shield-rotation.html debug page for an example...

**Comment by DavidBuerer** [comment]
@wipfli added contentMatch.html

**Comment by wipfli** [comment]
Thanks for the debug page. It looks like this:

image

Here, the top image has `contentMatch: 'height',` while the bottom has nothing.

**Comment by zekefarwell** [comment]
Term definitions for below:
- **Scale:** both x and y dimensions change keeping aspect ratio the same
    - Scale up: both dimensions increase
    - Scale down: both dimensions decrease
- **Stretch:** x dimension increases while y stays the same, or y dimension increases while x stays the same
- **Squeeze:** x dimension decreases while y stays the same, or y dimension decreases while x stays the same

It seems like the goal is to be able to to specify that an icon should _scale_ up or _scale_ down to fit the content (text) in one dimension, but then _stretch_ as needed in the other dimension, but not _squeeze_.  So two property values something like this:

- `scaleYstretchX`:  _Scale_ icon to content height, then _stretch_ icon width to content width if needed. Don't _squeeze_.
- `scaleXstretchY`:  _Scale_ icon to content width, then _stretch_ icon height to content height if needed. Don't _squeeze_.

Does this sum up the intent accurately, or am I still misunderstanding?

**Comment by HarelM** [comment]
The definitions you provided here are interesting, and helpful I believe, but then the definition of `scaleXstretchY` is confusing as you defined scale as an increase in both directions, so what does scaleX means which is different than stretchX?

If you were writing something like:
`ScaleAndStrechXOnly`, `stretchXScale`, `ScaleStretchX` I would say it might be more accurate, but it's still is not well defined what to expect from `ScaleAndStrechXOnly` to actually do... IDK...

**Comment by DavidBuerer** [comment]
@wipfli Not overshrinking is the goal, but how it overshrinks depends on the axis that you want to limit shrinking.  It is also how we are limiting the overshrinking.  There are other ways that it could be limited besides using aspect ratio.  Clarifying which technique allows for future expansion.

@zekefarwell Stretch doesn't imply increasing in that the stretchX area can increase or decrease depending on the size of the original icon and the size of the content.  I could see an argument that it should and that stretchX should be scaleX, but a little late for that.

**Comment by DavidBuerer** [comment]
@zekefarwell Yes, your diagram essentially describes contentMatch=height.

**Comment by DavidBuerer** [comment]
In case it wasn't aptly communicated, regardless of the solution, having the ability to do this will mean you only need one shield in cases like this.  Since each of these images is copied per pixel, there are startup performance savings.
image
https://api.maptiler.com/maps/streets/sprite@2x.png

**Comment by wipfli** [comment]
Here is a general discussion how we can manage controversial pull requests: https://github.com/maplibre/maplibre/discussions/155

**Comment by birkskyum** [comment]
Additional name proposal: flexibleContentRectangle

**Comment by DavidBuerer** [comment]
These are the proposed options so far:

1. contentMatch { none, height, width } (match content rectangle against specified axis and restrict it from shrinking below aspect ratio)
2. contentFitLimit { none, xByAspectRatio, yByAspectRatio, byAspectRatio } (content rectangle will use the specified instruction when fitting)
3. min-width and min-width-scale (limit the content rectangle in original image space and apply those limits when shrinking down to content)

@zekefarwell I'm not sure how you want to adapt your suggestion.
@birkskyum Your suggestion of flexibleContentRectangle implies to me that it would be an alternate rectangle to the content rectangle--which raises a ton of questions of which one takes precedence, what does that rectangle do exactly, etc.

**Comment by birkskyum** [comment]
Here is a poll https://github.com/maplibre/maplibre-gl-js/discussions/1944

**Comment by HarelM** [comment]
We should wait with the poll to get @zekefarwell input and incorporate it. I would feel better if we have 5 options or more to select from.

**Comment by wipfli** [comment]
Here are my two suggestions:

`scaleShield: none, horizontally, vertically`

`scaleSprite: none, horizontally, vertically`

**Comment by wipfli** [comment]
@DavidBuerer keep in mind that I don't even know what a sprite is, so maybe my suggestions are bad, but the problem this pull request solves seems to simply be to scale shields horzontally...

**Comment by wipfli** [comment]
Ah or we could call it `autoscaleShield` which to me implies a bit of magic

**Comment by DavidBuerer** [comment]
@wipfli It limits the sprite from scaling too much.  It must still be possible to stretch/shrink it normally if that's what you want.

**Comment by wipfli** [comment]
We could call it `doNotOvershrink: {none, horizontally, vertivally}`

**Comment by zekefarwell** [comment]
Since the idea here is to modify the behavior of the `icon-text-fit` layout property, it makes sense to me that the new property would be named similarly.  Here are the `icon-text-fit` values and what they appear to do (I've expanded a bit from the documentation):
- `none` - Display the icon at its intrinsic size and aspect ratio.
- `width` - Match the icon width to the text width by stretching (increasing) or shrinking (decreasing) icon width.
- `height` - Match the icon height to the text height by stretching or shrinking icon height.
- `both` - Match the icon width and height to the text width and height by stretching or shrinking both dimensions.

With the above in mind I would propose two new optional Sprite properties that can modify the behavior:

> `textFitWidth` - Specifies the scaling method used to modify the icon width when icon-text-fit: width|both is set.
> Values:
> - `stretchOrShrink` - (default) Match icon width to text width by stretching or shrinking icon width.  
> - `stretchOnly` - Stretch (increase) icon width to text width only if text width is larger.    Do not shrink (decrease) width.
> - `proportional` -  Scale icon width and height proportionally up or down so that icon width matches text width.
> 
> `textFitHeight` - Specifies the scaling method used to modify the icon height when icon-text-fit: height|both is set.
> Values:
> - `stretchOrShrink` - (default) Match icon height to text height by stretching or shrinking icon height.
> - `stretchOnly` - Stretch (increase) icon height to text height only if text height is larger.    Do not shrink (decrease) height.
> - `proportional` - Scale icon width and height proportionally up or down so that icon height matches text height.

To match the icon to the text height, but not over shrink the width, you would set:
```
textFitHeight: proportional
textFitWidth: stretchOnly
```
The inverse would match the icon to text width, but not over shrink the height.

If you wanted to fit text with the icon with fully proportionally scaling (no distortion) you would set:
```
textFitHeight: proportional
textFitWidth: proportional
```

If you wanted the icon to stretch in either dimension when the text is too large, but never shrink below the original icon size you would set:
```
textFitHeight: stretchOnly
textFitWidth: stretchOnly
``` 

The `stretchOrShrink` value could perhaps be left out since it has the same effect as the not including the property at all, but it felt clearer list it here to explain the different options.

It actually feels a bit weird to me for these to be Sprite properties since they modify how a Layout property behaves, but perhaps it makes sense since the `stretchX` and `stretchY` Sprite properties do that already.

**Comment by HarelM** [comment]
@zekefarwell thanks for the suggestion!
From what you wrote it feels like there's a missing option, I'm not sure if there's a valid use case for it, but if you have `stretchOnly` I would expect to have `shrinkOnly`, I'm not sure why, but it feels like it would be needed for completeness I guess... IDK.
thanks for the detailed comment!

**Comment by zekefarwell** [comment]
> it feels like there's a missing option, I'm not sure if there's a valid use case for it, but if you have `stretchOnly` I would expect to have `shrinkOnly`,

That does make sense.  I didn't include it because I couldn't think of a use case for it, maybe someone else can.  `shrinkOnly` would mean you want to shrink the icon size down to the text size when the text is smaller, but if the text is bigger than the icon you want to leave the icon at the original size with the text hanging out over the sides.  Seems kind of odd, but maybe there is some use case where this makes sense.

**Comment by DavidBuerer** [comment]
@zekefarwell The reason I put it as a sprite property is that you can have multiple icons in a layer and they may not have the same stretch characteristics.

**Comment by jonahadkins** [comment]
great addition to the codebase @DavidBuerer 👏   - coming in late here, but after reading through i'd lean toward `contentMatch { none, height, width }` option

**Comment by HarelM** [comment]
Please vote here:
https://github.com/maplibre/maplibre-gl-js/discussions/1944
Vote will be closed by the next TSC meeting in a week (Jan-11-2023)

**Comment by drwestco** [comment]
This example is confusing. From the description, both values affect the icon width. Which one wins?

```
textFitHeight: proportional
textFitWidth: stretchOnly
```

**Comment by zekefarwell** [comment]
> This example is confusing. From the description, both values affect the icon width. Which one wins?

Sorry, I didn't explain this in the proposal, but there would need to be a specific order of operations that evaluates any `proportional` value before the others.  In this case, `textFitHeight: proportional`, would be evaluated first, scaling both dimensions proportionally until the icon height matches the text height.  Then `textFitWidth: stretchOnly` would be evaluated after to further stretch the width if necessary.  Here's a visual:
203826232-51dabfda-4353-4c4a-baf7-143b7b124e75

The actual image scaling could happen in one step by doing the math in two steps ahead of time.

**Comment by xabbu42** [comment]
The original proposal had three settings, one for the current behavior and two new ones. I get that the two newly solved use cases are shields with vertical text where the height should not get too small and shields with horizontal texts where the width should not get too small. The proposal from @zekefarwell has a total of 9 different settings. What are the use cases for the additional 6 settings?

**Comment by wipfli** [comment]
@DavidBuerer what is the status on this pull request?

**Comment by DavidBuerer** [comment]
> 

I've been focused on other things and haven't had time to invest in it... especially in a rewrite.

**Comment by HarelM** [comment]
Conflicts piling up...
Please let us know if this is something you would like to push forward, otherwise I'll close this PR.

**Comment by DavidBuerer** [comment]
> Conflicts piling up... Please let us know if this is something you would like to push forward, otherwise I'll close this PR.

We rely on this change internally, so it would be good to get some flavor of it integrated so that we don't have to resolve merge conflicts.  I didn't see a response to the comment by @xabbu42 but it seems like the most popular implementation is textFitWidth/Height by @zekefarwell?  I'll see if I can carve out some time to implement that.

**Comment by HarelM** [comment]
@DavidBuerer @drwestco I'm closing this PR.
Feel free to open a new one if you would like to push this forward.
I've completely refactored all the callback hell in maplibre, so I hope rewriting this would be much simpler now if this is still something you would like to push forward.

**Comment by HarelM** [comment]
For anyone subscribed to this thread, the relevant linked PR was just merged.
